src/serialize.rs

```diff
@@ -1,67 +1,76 @@
 use std::io::Cursor;
 use std::io::Read;
-use std::io::Write;
+use std::io::{Error, ErrorKind};
 
 use crate::allocator::Allocator;
 use crate::node::Node;
 
 const MAX_SINGLE_BYTE: u8 = 0x7f;
 const CONS_BOX_MARKER: u8 = 0xff;
 
-fn encode_size(f: &mut dyn Write, size: usize) -> std::io::Result<()> {
+fn bad_encoding() -> std::io::Error {
+    Error::new(ErrorKind::InvalidInput, "bad encoding")
+}
+
+fn encode_size(out: &mut Vec<u8>, size: usize) {
     if size < 0x40 {
-        f.write_all(&[(0x80 | size) as u8])?;
+        out.push((0x80 | size) as u8);
     } else if size < 0x2000 {
-        f.write_all(&[(0xc0 | (size >> 8)) as u8, ((size) & 0xff) as u8])?;
+        out.extend(&[(0xc0 | (size >> 8)) as u8, ((size) & 0xff) as u8]);
     } else if size < 0x10_0000 {
-        f.write_all(&[
+        out.extend(&[
             (0xe0 | (size >> 16)) as u8,
             ((size >> 8) & 0xff) as u8,
             ((size) & 0xff) as u8,
-        ])?;
+        ]);
     } else if size < 0x800_0000 {
-        f.write_all(&[
+        out.extend(&[
             (0xf0 | (size >> 24)) as u8,
             ((size >> 16) & 0xff) as u8,
             ((size >> 8) & 0xff) as u8,
             ((size) & 0xff) as u8,
-        ])?;
+        ]);
     } else if size < 0x4_0000_0000 {
-        f.write_all(&[
+        out.extend(&[
             (0xf8 | (size >> 32)) as u8,
             ((size >> 24) & 0xff) as u8,
             ((size >> 16) & 0xff) as u8,
             ((size >> 8) & 0xff) as u8,
             ((size) & 0xff) as u8,
-        ])?;
+        ]);
+    } else {
+        panic!("atom size exceeded maximum {}", size);
     }
-    Ok(())
 }
 
-pub fn node_to_stream<T: Allocator>(node: &Node<T>, f: &mut dyn Write) -> std::io::Result<()> {
+pub fn node_to_bytes_helper<T: Allocator>(node: &Node<T>, out: &mut Vec<u8>) {
```

Comment:
I'm not a big fan of this change, as writing to a stream is more general than writing to memory. I know you're just trying to get rid of the error case, but we can do that at a higher level (ie. with `.unwrap`) if we know we're writing to memory and not likely to fail.

Reply:
I don't have a strong opinion about it, but we don't need it to be more general. Do you foresee that we will need to write it to something other than memory?

Reply:
I don't see a specific need for the near term... it just makes my Spidey senses tingle.